tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,310 @@
+            .unwrap_or_else(|| {
+                panic!("tried to drop {:?} but it no longer exists!", id);
+            })
+        {
+            return false;
+        }
+
+        // Synchronize only if we are actually removing the span (stolen
+        // from std::Arc);
+        fence(Ordering::Acquire);
+        self.spans.remove(idx)
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let mut values = HashMap::new();
+        let mut visitor = RegistryVisitor(&mut values);
+        attrs.record(&mut visitor);
+        let parent = attrs.parent().map(|id| id.clone());
+        let s = Data {
+            metadata: attrs.metadata(),
+            parent,
+            children: vec![],
+            ref_count: AtomicUsize::new(1),
+            values: Mutex::new(values),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    #[inline]
+    fn record(&self, span: &span::Id, values: &span::Record<'_>) {
+        dbg!(span);
+        dbg!(values);
+    }
+
+    fn record_follows_from(&self, span: &span::Id, follows: &span::Id) {
+        let current = self
+            .get(span)
+            .expect("Current span is missing; this is a bug");
+
+        let current = self
+            .span(span)
+            .expect("Current span is missing; this is a bug");
+    }
+
+    fn enter(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
+            spans.borrow_mut().push(id.clone());
+            if let Some(data) = self.get(id) {
+                data.ref_count.fetch_add(1, Ordering::SeqCst);
+            } else {
+                panic!("Span with ID {:?} was not found. This is a bug.", id);
+            }
+        })
+    }
+
+    fn event(&self, event: &Event<'_>) {
+        let id = match event.parent() {
+            Some(id) => Some(id.clone()),
+            None => {
+                if event.is_contextual() {
+                    CURRENT_SPANS.with(|spans| spans.borrow().current().map(|id| id.clone()))
+                } else {
+                    None
+                }
+            }
+        };
+        if let Some(id) = id {
+            let mut values = HashMap::new();
+            let mut visitor = RegistryVisitor(&mut values);
+            event.record(&mut visitor);
+            let _ = self.get(&id).expect("Missing parent span for event");
+        }
+    }
+
+    fn exit(&self, id: &span::Id) {
+        CURRENT_SPANS.with(|spans| {
+            spans.borrow_mut().pop(id);
+            if let Some(data) = self.get(id) {
```

Comment:
Couple of things:
- Why not use the existing `drop_span` function rather than reimplementing _some_ of its logic?
- Why sequentially consistent?

Also, in order to ensure close notifications are propagated to layers, this needs to go through the dispatcher. Something like:
```rust
fn exit(&self, id: &span::Id) {
    let id = CURRENT_SPANS.with(|spans| spans.borrow_mut().pop(id));
    dispatcher::get_current(|current| current.try_close(id));
}
```

Reply:
Made this instead:

```rust
fn exit(&self, id: &span::Id) {
    let _ = CURRENT_SPANS.with(|spans| spans.borrow_mut().pop(id));
    dispatcher::get_default(|dispatch| dispatch.try_close(id.clone()));
}
```